server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java

```diff
@@ -0,0 +1,435 @@
+         * until the transmission of the chunk that is currently under construction.
+         */
+        private void transferCurrentFragmentReleasable(ArrayList<Releasable> releasables) {
+            assert queueRefs.hasReferences();
+
+            if (currentFragmentReleasable == null) {
+                return;
+            }
+
+            if (releasables == nextReleasablesCache) {
+                // adding the first value, so we must line up a new cached value for the next caller
+                nextReleasablesCache = new ArrayList<>();
+            }
+
+            releasables.add(currentFragmentReleasable);
+            currentFragmentReleasable = null;
+        }
+
+        @Override
+        public ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) throws IOException {
+            assert Transports.isTransportThread(Thread.currentThread());
+
+            final ArrayList<Releasable> releasables = nextReleasablesCache;
+            assert releasables.isEmpty();
+            try {
+                if (tryAcquireQueueRef()) {
+                    try {
+                        assert queueLength.get() > 0;
+                        // This is the current unique active AvailableFragmentsResponseBodyPart (i.e. queueLength is strictly positive and
+                        // we hold a queueRef), so any concurrent calls to writeFragment() at this point will just add to the queue and
+                        // won't spawn a new AvailableFragmentsResponseBodyPart or mutate any fields.
+
+                        final RecyclerBytesStreamOutput chunkStream = new RecyclerBytesStreamOutput(recycler);
+                        assert targetStream == null;
+                        targetStream = chunkStream;
+
+                        do {
+                            if (fragmentChunksIterator.hasNext()) {
+                                fragmentChunksIterator.next().toXContent(xContentBuilder, params);
+                            } else {
+                                completeCurrentFragment(releasables);
+                            }
+                        } while (isResponseComplete == false && isResponsePaused == false && chunkStream.size() < sizeHint);
+
+                        assert (releasables == nextReleasablesCache) == releasables.isEmpty();
+                        assert nextReleasablesCache.isEmpty();
+
+                        final Releasable chunkStreamReleasable = () -> Releasables.closeExpectNoException(chunkStream);
+                        final var result = new ReleasableBytesReference(
+                            chunkStream.bytes(),
+                            releasables.isEmpty()
+                                ? chunkStreamReleasable
+                                : Releasables.wrap(Iterators.concat(Iterators.single(chunkStreamReleasable), releasables.iterator()))
+                        );
+                        targetStream = null;
```

Comment:
How about splitting up these lines into a new method `createReleasableBytesReference`?

Reply:
Again, I'm not seeing the benefit.